hex/spec/dependabot/hex/update_checker_spec.rb

```diff
@@ -359,6 +359,104 @@
       end
     end
 
+    context "with a dependency from a private repo" do
+      let(:mixfile_body) { fixture("mixfiles", "private_repo") }
+      let(:lockfile_body) { fixture("lockfiles", "private_repo") }
+
+      before { `mix hex.repo remove dependabot` }
+
+      let(:dependency_name) { "jason" }
+      let(:version) { "1.0.0" }
+      let(:dependency_requirements) do
+        [{ file: "mix.exs", requirement: "~> 1.0.0", groups: [], source: nil }]
+      end
+
+      context "with good credentials" do
+        let(:credentials) do
+          [{
+            "type" => "hex_repository",
+            "repo" => "dependabot",
+            "auth_key" => "d6fc2b6n6h7katic6vuq6k5e2csahcm4",
+            "url" => "https://dependabot-private.fly.dev"
+          }]
+        end
+
+        it { is_expected.to eq(Dependabot::Hex::Version.new("1.1.0")) }
+      end
+
+      context "with bad credentials" do
+        let(:credentials) do
+          [{
+            "type" => "hex_repository",
+            "repo" => "dependabot",
+            "auth_key" => "111f6cbeffc6e14c6a884f0111caff3e",
+            "url" => "https://dependabot-private.fly.dev"
+          }]
+        end
+
+        it "raises a helpful error" do
+          error_class = Dependabot::PrivateSourceAuthenticationFailure
+
+          expect { subject }.
+            to raise_error(error_class) do |error|
+              expect(error.source).to eq("dependabot")
+            end
+        end
```

Comment:
😞  I'll add a unit test to `dependabot-private-repo` to verify auth and post it back here.

Reply:
It looks like the server is validating with the token as expected:

```
$ curl -I -H "Authorization: d6fc2b6n6h7katic6vuq6k5e2csahcm4" https://dependabot-private.fly.dev/versions
HTTP/2 200
accept-ranges: bytes
cache-control: public
content-length: 319
content-type: application/octet-stream
date: Thu, 10 Nov 2022 15:44:37 GMT
etag: "2606F92"
server: Fly/b1863e2e7 (2022-11-09)
via: 2 fly.io
fly-request-id: 01GHH3DW2QSQQHC1RV14QZYF88-chi

$ curl -I -H "Authorization: not-the-real-token" https://dependabot-private.fly.dev/versions
HTTP/2 401
cache-control: max-age=0, private, must-revalidate
content-length: 32
date: Thu, 10 Nov 2022 15:44:55 GMT
server: Fly/b1863e2e7 (2022-11-09)
via: 2 fly.io
fly-request-id: 01GHH3ED8671MHVV67SZ0GZ8YW-chi

$ curl -I -H "Authorization: not-the-real-token" https://dependabot-private.fly.dev/public_key
HTTP/2 200
accept-ranges: bytes
cache-control: public
content-length: 427
content-type: application/octet-stream
date: Thu, 10 Nov 2022 15:45:48 GMT
etag: "7CEF702"
server: Fly/b1863e2e7 (2022-11-09)
via: 2 fly.io
fly-request-id: 01GHH3G0VM1C56N43CG7EPZX7P-chi
```

I'm investigating the checker spec and code now.

Reply:
Ah, it's because it is fetching the package from a local cache:

```
#<Dependabot::SharedHelpers::HelperSubprocessFailed: Failed to fetch record for 'dependabot/jason' from registry (using cache instead)
```

It's a legitimate bug in this branch because the "Failed to fetch record" regex only checks for `hexpm` strings!

Reply:
@landongrindheim Failure fixed with b92d1043f 👍